### PR TITLE
dependency updates

### LINKED ISSUE #538: MPI AMIP occasional hang
Our MPI AMIP run is hanging almost every time it's run.

It looks like the error originated in the PR where we updated dependencies https://github.com/CliMA/ClimaCoupler.jl/pull/535, which updated from ClimaCore v0.10 to v0.11 and also removed a pin for HDF5 at 1.12. That branch had the first hangs, though some commits hung and others passed. This branch was merged on 12/6.

Hangs on [this line](https://github.com/CliMA/ClimaCoupler.jl/blob/main/src/Regridder.jl#L335), where `ClimaCore.InputOutput.read_field` is called.

Edit: We've discovered that the issue is coming from memoizing Grids using a WeakValueDict, which was introduced in ClimaCore v0.11.0 See ClimaCore issue [#1589](https://github.com/CliMA/ClimaCore.jl/issues/1589) for more details. A fix will need to be implemented in ClimaCore.

### Examples of failure on buildkite
- [debug branch 2298 (has useful print statements)](https://buildkite.com/clima/climacoupler-ci/builds/2298#018cd097-0dbe-4d22-bc9d-dc1333139e2c)
- [dep updates PR 12/5 2208](https://buildkite.com/clima/climacoupler-ci/builds/2208)
- [dep updates PR 12/6 2223](https://buildkite.com/clima/climacoupler-ci/builds/2223#018c40b2-d596-4bbf-ae4e-c57216a67182)
- [dep updates PR 12/6 2220](https://buildkite.com/clima/climacoupler-ci/builds/2220)
- [on main 12/7 2229](https://buildkite.com/clima/climacoupler-ci/builds/2229#018c41cb-fba6-4b5b-b7ef-1660dd6e7aaa)
- [depot branch 12/8 2237](https://buildkite.com/clima/climacoupler-ci/builds/2237#_)
- this run [passed](https://buildkite.com/clima/climacoupler-ci/builds/2281), then [failed](https://buildkite.com/clima/climacoupler-ci/builds/2299) when re-run without any changes to the code.
- same error in MPI BCReader and MPI Regridder tests, [debug branch 1/4 2311](https://buildkite.com/clima/climacoupler-ci/builds/2311#_)

### Debugging notes
- increased `slurm_mem` for MPI AMIP job from 16GB to 32GB
- updated pipeline to match ClimaCore's: increased `JULIA_MAX_NUM_PRECOMPILE_FILES` from 50 to 100; added `slurm_mem: 80G` to init environment step; removed `SLURM_KILL_BAD_EXIT: 1`
- exits with error 137 (out of memory), but increasing memory from 16GB to 32GB didn't prevent hang
- downgraded ClimaCore 0.11.7 -> 0.11.3 (still hangs)
- downgraded ClimaCore 0.11.7 -> 0.10.56, ClimaAtmos 0.17 -> 0.16.3, Insolation 0.80 -> 0.6.0, RRTMGP -> 0.9.3
  - (I just wanted to downgrade ClimaCore but had to do them all)
  - this [passed](https://buildkite.com/clima/climacoupler-ci/builds/2335)!!!!!!!! the breaking change may have been made in ClimaCore v0.11.0 - 0.11.3. The only HDF5-related changes were made in 0.11.0 so it's probably something there

### LINKED ISSUE #547: update to julia v1.10
Julia v1.10 is now available, and we should update our code to use it. We may need to wait until upstream packages update their compatibility for v1.10 before we update.

#432 updated julia v1.8 -> v1.9 (we can use as a template)

### other repos
- ClimaLSM update: https://github.com/CliMA/ClimaLSM.jl/issues/437

### LINKED ISSUE #554: dependency updates
update to ClimaAtmos v0.18, ClimaLSM 0.5

closed by #556

### notes
- merge after #538 is fixed
- ClimaLSM v0.5.0 is incompatible with NCDatasets v0.14, so we'll have to update NCDatasets in another PR after LSM releases next

### PR DESCRIPTION
**Pull request by juliasloan25**
## Purpose 
update to ClimaAtmos v0.19.0, ClimaLSM v0.5.0. also updates to julia v1.10

wait to merge until #538 is fixed (see https://github.com/CliMA/ClimaCore.jl/issues/1589)

closes #554 (dependency updates)
closes #547 (julia v1.10 update)
closes #538 (MPI hang)
supersedes #549 
supersedes #550 
supersedes #558 

## To-do
- [x] ClimaLSM breaking changes
- [x] ClimaAtmos breaking changes (none)
- [x] change branch protection rule for main: ci 1.9 -> ci 1.10

### CODE REVIEW

File: experiments/AMIP/components/land/bucket_init.jl [hunk 1]
```diff
@@ -135,25 +122,41 @@ function surface_air_density(
...
+    domain_names = ([:surface for k in keys]...,)
+    model_name = :drivers
+    # intialize_vars packages the variables as a named tuple,
+    # as part of a named tuple with `model_name` as the key.
+    # Here we just want the variable named tuple itself
+    vars = ClimaLSM.initialize_vars(keys, types, domain_names, coords, model_name)
+    return vars.drivers
+end
+
+function ClimaLSM.initialize_drivers(a::CoupledRadiativeFluxes{FT}, coords) where {FT}
```
Comment (juliasloan25) [inline experiments/AMIP/components/land/bucket_init.jl:158]: todo: remove this function after ClimaLSM v0.6.0 release + update. Maybe we can actually add both these initialize_drivers methods to ClimaLSM? @kmdeck

File: .buildkite/longruns/pipeline.yml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 agents:
   queue: central
   slurm_time: 24:00:00
-  modules: julia/1.9.4 cuda/12.2 ucx/1.14.1_cuda-12.2 openmpi/4.1.5_cuda-12.2 nsight-systems/2023.3.1
+  modules: julia/1.10.0 cuda/12.2 ucx/1.14.1_cuda-12.2 openmpi/4.1.5_cuda-12.2 nsight-systems/2023.3.1
```
Comment (Sbozzolo) [inline .buildkite/longruns/pipeline.yml:4]: You can also upgrade to nsight 2023.4.1

File: .buildkite/longruns/pipeline.yml [hunk 3]
```diff
@@ -34,7 +34,7 @@ steps:
       # TODO: This can be removed once all the branches that use MPIPreferences
       # are deleted
       - echo "--- Remove MPIPreferences"
-      - "rm -f ${JULIA_DEPOT_PATH}/environments/v1.9/LocalPreferences.toml"
+      - "rm -f ${JULIA_DEPOT_PATH}/environments/v1.10/LocalPreferences.toml"
```
Comment (Sbozzolo) [inline .buildkite/longruns/pipeline.yml:37]: You can remove this section.

The `rm` was there while there were unupdated branches, but nothing will write the wrong entries to `v1.10/LocalPreferences.toml`.

File: Project.toml [hunk 4]
```diff
@@ -25,11 +25,11 @@ Thermodynamics = "b60c26fb-14c3-4610-9d3e-2d17fe7ff00c"
 
 [compat]
 CLIMAParameters = "0.4, 0.5, 0.6, 0.7"
-ClimaAtmos = "0.17.1"
+ClimaAtmos = "0.18"
```
Comment (Sbozzolo) [inline Project.toml:28]: Maybe you can move to 0.19 directly?

I don't think there's going to be any challenge.

File: test/mpi_tests/local_checks.sh [hunk 5]
```diff
@@ -7,7 +7,7 @@
 #SBATCH --ntasks=2
 
 module purge
-module load julia/1.9.4 openmpi/4.1.1 hdf5/1.12.1-ompi411
+module load julia/1.10.0 openmpi/4.1.1 hdf5/1.12.1-ompi411
```
Comment (Sbozzolo) [inline test/mpi_tests/local_checks.sh:10]: This is outdated

File: experiments/AMIP/components/land/bucket_init.jl (hunk 1 above)
Comment (kmdeck) [reply experiments/AMIP/components/land/bucket_init.jl:158]: yes! I will start on moving them over this week & ask for review. I am thinking the struct definitions of the `Coupled` types would also move?

Comment (juliasloan25) [reply experiments/AMIP/components/land/bucket_init.jl:158]: Yeah, I think they should be moved to src/

**Review by Sbozzolo — APPROVED** [review]
Looks good!

**Pull request merged**